appcues/src/main/java/com/appcues/ui/composables/AppcuesComposition.kt

```diff
@@ -112,16 +109,50 @@ private fun BoxScope.ComposeLastRenderingState(state: Rendering) {
     ComposeContainer(state.stepContainer, state.position)
 }
 
+private suspend fun produceMetadata(
+    metadataSettingTraits: List<MetadataSettingTrait>,
+    traitExceptionHandler: AppcuesTraitExceptionHandler,
+    overlayVisibility: AppcuesOverlayVisibility,
+): HashMap<String, Any?>? {
+    return try {
+        hashMapOf<String, Any?>().apply { metadataSettingTraits.forEach { putAll(it.produceMetadata()) } }.also {
+            overlayVisibility.setVisible(true)
+        }
+    } catch (ex: AppcuesTraitException) {
+        if (ex.retryMilliseconds != null) {
+            overlayVisibility.setVisible(false)
+            delay(ex.retryMilliseconds.toLong())
+            produceMetadata(metadataSettingTraits, traitExceptionHandler, overlayVisibility)
+        } else {
+            traitExceptionHandler.onTraitException(ex)
+            null
+        }
+    }
+}
+
 @Composable
 internal fun BoxScope.ComposeContainer(stepContainer: StepContainer, stepIndex: Int) {
     with(stepContainer) {
         val backdropDecoratingTraits = remember(stepIndex) { mutableStateOf(steps[stepIndex].backdropDecoratingTraits) }
         val containerDecoratingTraits = remember(stepIndex) { mutableStateOf(steps[stepIndex].containerDecoratingTraits) }
         val metadataSettingTraits = remember(stepIndex) { mutableStateOf(steps[stepIndex].metadataSettingTraits) }
-        val stepMetadata = rememberMetadataSettingTraits(metadataSettingTraits)
 
-        if (stepMetadata != null) {
-            CompositionLocalProvider(LocalAppcuesStepMetadata provides stepMetadata.value) {
+        val traitExceptionHandler = LocalAppcuesTraitExceptionHandler.current
+        val overlayVisibility = LocalAppcuesOverlayVisibility.current
+        val stepMetadata = remember { mutableStateOf<AppcuesStepMetadata?>(null) }
+        val previousStepMetaData = remember { mutableStateOf(AppcuesStepMetadata()) }
+
+        LaunchedEffect(metadataSettingTraits) {
+            val actual = produceMetadata(metadataSettingTraits.value, traitExceptionHandler, overlayVisibility)
+            if (actual != null) {
+                stepMetadata.value = AppcuesStepMetadata(previous = previousStepMetaData.value.current, current = actual).also {
+                    previousStepMetaData.value = it
+                }
+            }
+        }
```

Comment:
please review this section carefully around he usage of the `LaunchedEffect` and how it is replacing what was previously in `rememberMetadataSettingsTraits` with the new call to the suspend fun for `produceMetadata(...)`. This allows for the async retry within `produceMetadata` and then the rest of the composition relies on `stepMetadata.value` (L154 below) to continue composing once valid metadata is produced (or fails through exception handler if not).

Mostly, I want to make sure I didn't miss any Compose gotchas around how that previous remember was working, which could cause any unexpected side effects or recomposition issues.

Reply:
ok my understanding is that whenever metadataSettingTraits changes (so on a new stepIndex) then try to produce metadata and set as the value, this looks ok change